ompi/mca/mtl/psm2/mtl_psm2.c

```diff
@@ -406,58 +406,62 @@ int ompi_mtl_psm2_progress( void ) {
     int completed = 1;
 
     do {
+        OPAL_THREAD_LOCK(&mtl_psm2_mq_mutex);
         err = psm2_mq_ipeek2(ompi_mtl_psm2.mq, &req, NULL);
-	if (err == PSM2_MQ_INCOMPLETE) {
-	    return completed;
-	} else if (err != PSM2_OK) {
-	    goto error;
-	}
+        if (err == PSM2_MQ_INCOMPLETE) {
```

Comment:
is this likely or unlikely? or does it matter? I noted you used opal_unlikely below, hence the question

Reply:
mm... both should happen in every call. The loop will iterate until emptying PSM2's  Matched Queue.  How many times will go on each direction will depend on each app, e.g. in flight yet to be poled/tested requests.

Reply:
oh, the unlikely is for psm2_mq_test2() failing to on a psm request that psm2_mq_ipeek2() returned as ready to be tested. That should never fail, but..